Simplified error messages on sign in.

tclaus opened the pull request:
Changed technical messages to generic message.
In iOS case it could be possible, that Main view occured, but with a error

Changed from 
<img width="375" alt="Bildschirmfoto 2020-11-06 um 18 27 40" src="https://user-images.githubusercontent.com/501326/98404036-2c7ddb80-206a-11eb-8d70-ced1456504dd.png">

To: 
<img width="375" alt="Bildschirmfoto 2020-11-06 um 19 33 57" src="https://user-images.githubusercontent.com/501326/98404053-33a4e980-206a-11eb-9fd5-a0e8daaadcae.png">

tclaus committed:
Simplified error messages on sign in.
Changed technical messages to generic message.
In iOS case it could be possible, that Main view occured, but with a error

jhass commented:
I feel this too simplistic and often outright wrong. There's many more error cases in here than "could not resolve host". You also simply dropped the logic recovering from an expired or removed authorization for example. Also what is an "Invalid API", I don't think that's a good name :)

jhass commented on lib/src/app_auth.dart:
```diff
@@ -318,3 +313,11 @@ class InvalidSessionError implements Exception {
+
+/**
+ * API is invalid
+ */
```
In general I feel like it's better to omit such overly generic comments. There's no additional information here that's not already conveyed in the class name.

jhass commented on lib/src/localizations.dart:
```diff
@@ -156,6 +156,13 @@ class InsporationLocalizations {
+      name: 'errorSignInNoHost',
+      desc: 'Error message if no host can not be resolved',
+      locale: localeName
+  );
```
When you add a new string, please run `script/generate_template_locale` and commit the resulting changes.

jhass replied:
Also when an existing translation is edited :)

jhass commented on lib/src/localizations.dart:
```diff
@@ -156,6 +156,13 @@ class InsporationLocalizations {
   );
 
+  String get errorSignInNoHost => Intl.message(
+      'Unable to resolve host, are you sure this pod supports the API?',
```
The message seems misleading. Not being able to resolve the hostname tells us nothing about whether the pod supports the API or not.

tclaus replied:
Good point. 
Should we insert an extra layer before auth process to check if service supports the api?
Maybe after "detect configuration" and in case of a error?

jhass replied:
How would you detect it?

tclaus replied:
🤔
What do you think to pin this on 'Discover service'? On new or unknown services, this is the first call. If we receive no answer, server is unavailable or doesn't know this service at all (invalidate API)

jhass replied:
But what error conditions would  lead to this invalidation?

The one you picked here (while burying many others under this message) is not indicative of the API being available at the pod or not at all.

tclaus commented:
It turns out, that on iOS side the exceptions thrown in case of an error don't follow the exception codes in the DART / KT file. I'll fix this

tclaus committed:
Updated Code in swift exceptions
Added timeout detection in iOS / swift side
Simplified error user-faciong texts

jhass commented on ios/Runner/AppAuthHandler.swift:
```diff
@@ -25,31 +29,53 @@ class AppAuthHandler {
+        predefineURLSession()
     }
     
+// Set some custom configurations for authorization sessions
```
```suggestion
    // Set some custom configurations for authorization sessions
```

jhass commented on ios/Runner/AppAuthHandler.swift:
```diff
@@ -83,15 +109,15 @@ class AppAuthHandler {
                 
                 if let error = error {
-                    self.invokeErrorHandler(errorMessage: error.localizedDescription)
+                    self.invokeErrorHandler(code: self.FAILED_TOKEN_FETCH, errorMessage: error.localizedDescription)
```
Is the `self` necessary here? Ditto all the others or missing above.

tclaus replied:
It's needed for closures to reference method explicit.

jhass replied:
But for the constant?

tclaus replied:
Thats swift.. 
<img width="834" alt="Bildschirmfoto 2020-11-07 um 22 30 27" src="https://user-images.githubusercontent.com/501326/98451764-27438e00-2149-11eb-991e-e80b451f2d88.png">

jhass commented on ios/Runner/AppAuthHandler.swift:
```diff
@@ -108,7 +134,7 @@ class AppAuthHandler {
         guard let userId = self.currentSession?.userId else {
             os_log("No userID set")
-            self.invokeErrorHandler(errorMessage: "No user ID set")
+            self.invokeErrorHandler(code: self.FAILED_AUTHORIZE, errorMessage: "No user ID set")
```
Ditto

jhass commented on ios/Runner/AppAuthHandler.swift:
```diff
@@ -117,20 +143,27 @@ class AppAuthHandler {
         OIDAuthorizationService.discoverConfiguration(forIssuer: discoveryURL) { (configuration, error) in
-            
-            if let error = error {
+            // 1001: Timeout, 1003 = Hostnot found
```
```suggestion
            // 1001: Timeout, 1003 = Host not found
```

Not sure what story this comment tells though, none of the detection code seems to make use of this fact.

tclaus replied:
One of my coding notes. Using system constants now.

jhass commented on ios/Runner/AppAuthHandler.swift:
```diff
@@ -177,9 +210,16 @@ class AppAuthHandler {
                 os_log("Failed registration client %{error}@", log: .default, type: .error, error.localizedDescription)
                 self.setAuthState(nil) // Clear local store - if already stored
-                self.invokeErrorHandler(errorMessage: "Failed to register service: \(error.localizedDescription)")
+                self.invokeErrorHandler(code: self.FAILED_REGISTER, errorMessage: "Failed to register service: \(error.localizedDescription)")
```
Is there a not "localized" description? Given we don't translate these (yet), mixing languages seems weird. Ditto all the others.

tclaus replied:
The 'localized' is not really localized, but user readable. We could make it accessible in a user facing message? To keep it in logs is more important. 
It could be hard to manual translate every kind of error - Will this help a user?

jhass replied:
It might help us to help a user.

tclaus replied:
Then the deeph should be fixed: On network level (After validating the username) there are three network calls: 
- Service discovery
- Register Client 
- Request Auth
- Get token

- Token Refresh

This should be discussed.

jhass replied:
Sorry, I cannot follow your comment.

jhass commented on ios/Runner/AppAuthHandler.swift:
```diff
@@ -281,7 +328,7 @@ extension AppAuthHandler {
         
         if let authState = state {
-            // Attention! This arhcive-Methis is marked as depricated, but recommended method
+            // Attention! This archivedData-method is marked as depricated in iOS 12.0, but recommended method
```
```suggestion
            // Attention! The archivedData method is marked as deprecated in iOS 12.0, but recommended method
```

jhass commented on ios/Runner/AppAuthHandler.swift:
```diff
@@ -300,7 +347,7 @@ extension AppAuthHandler {
         }
         
-        // Attention! This arhcive-Methis is marked as depricated, but recommended method
+        // Attention! This archivedData-method is marked as depricated, but recommended method
```
```suggestion
        // Attention! The archivedData method is marked as deprecated, but recommended method
```

jhass commented on lib/src/localizations.dart:
```diff
@@ -156,10 +156,10 @@ class InsporationLocalizations {
-      'Unable to resolve host, are you sure this pod supports the API?',
-      name: 'errorSignInNoHost',
-      desc: 'Error message if no host can not be resolved',
+  String get errorInvalidAPI => Intl.message(
```
```suggestion
  String errorAuthorizationFailed(String user) => Intl.message(
```

Adding the argument also requires `args` and `examples`.

jhass commented on lib/src/localizations.dart:
```diff
@@ -156,10 +156,10 @@ class InsporationLocalizations {
-      name: 'errorSignInNoHost',
-      desc: 'Error message if no host can not be resolved',
+  String get errorInvalidAPI => Intl.message(
+      'Service did not respond as expected. Is the hostname right? Does the pod run at least version 0.8?',
```
```suggestion
      'Could not authorize to $user, is it spelled correctly, is your network running and is your pod running the latest development snapshot?', // TODO update message after diaspora 0.8 release
```

jhass commented on lib/src/localizations.dart:
```diff
@@ -156,10 +156,10 @@ class InsporationLocalizations {
+  String get errorInvalidAPI => Intl.message(
+      'Service did not respond as expected. Is the hostname right? Does the pod run at least version 0.8?',
+      name: 'errorInvalidAPI',
+      desc: 'Error message host did not respond. Either invalid hostname or pod run does not at least on version 0.8 or higher',
```
```suggestion
      desc: 'Error message after authorization with the host failed',
```

jhass commented on lib/src/app_auth.dart:
```diff
@@ -318,3 +321,8 @@ class InvalidSessionError implements Exception {
   String toString() => "Invalid session: $message";
 }
+
+class InvalidAPIException implements Exception {
```
```suggestion
class AuthorizationFailedException implements Exception {
```

jhass commented on lib/src/app_auth.dart:
```diff
@@ -318,3 +321,8 @@ class InvalidSessionError implements Exception {
 }
+
+class InvalidAPIException implements Exception {
+  InvalidAPIException(this.message);
```
```suggestion
  AuthorizationFailedException(this.message);
```

jhass commented on lib/src/app_auth.dart:
```diff
@@ -108,17 +108,20 @@ class AppAuth {
-      // our session is probably not worth anything anymore, destroy it
+      if (e.code.startsWith("failed_")) {
+        // Catches all failed_discovery (Unknown service/API), failed_register, failed_auth. etc
+        throw InvalidAPIException(e.message);
```
```suggestion
        throw AuthorizationFailedException(e.message);
```

jhass commented on lib/src/app_auth.dart:
```diff
@@ -108,17 +108,20 @@ class AppAuth {
-        throw e.message; // probably bad network
-      }
 
+      // Check code before message
```
Well yes, obviously. The comment doesn't explain *why*

```suggestion
```

jhass commented on lib/src/app_auth.dart:
```diff
@@ -108,17 +108,20 @@ class AppAuth {
+        throw InvalidAPIException(e.message);
+      }
+
       await _destroyCurrentSession("Failed to fetch access token: ${e.message}");
```
```suggestion
      // our session is probably not worth anything anymore, destroy it
      await _destroyCurrentSession("Failed to fetch access token: ${e.message}");
```

jhass commented on lib/src/app_auth.dart:
```diff
@@ -108,17 +108,20 @@ class AppAuth {
+
       await _destroyCurrentSession("Failed to fetch access token: ${e.message}");
       return null; // Previous always raises
+
```
```suggestion
```

jhass commented on lib/sign_in_page.dart:
```diff
@@ -256,6 +256,11 @@ class _SignInPageState extends State<SignInPage> with StateLocalizationHelpers {
         _lastError = l.errorSignInTimeout;
         _loading = false;
       });
+    } on InvalidAPIException {
```
```suggestion
    } on AuthorizationFailedException {
```

jhass commented on lib/sign_in_page.dart:
```diff
@@ -256,6 +256,11 @@ class _SignInPageState extends State<SignInPage> with StateLocalizationHelpers {
       });
+    } on InvalidAPIException {
+      setState(() {
+        _lastError = l.errorInvalidAPI;
```
```suggestion
        _lastError = l.errorAuthorizationFailed(userId);
```

jhass reviewed (changes requested):
After this, how do we troubleshoot users failing to sign in? Maybe for the time being it would be good to still append the original exception message, despite not being able to localize it?

jhass commented on lib/src/app_auth.dart:
```diff
@@ -108,17 +108,20 @@ class AppAuth {
     } on PlatformException catch (e) {
-      if (e.message?.toLowerCase()?.contains("network") == true) {
-        throw e.message; // probably bad network
-      }
```
Shouldn't we keep this heuristic? Could map to another "nice" user facing message

tclaus replied:
To check the occurrence of a single word in a description text... is is named a 'heuristic' - could call it a 'guess'.- 
My proposal tries to use all known exceptions to use the codes: 'timeout'..'failed_...',  Here we have a chance to simplify messages without considering every possible network error. 

I tested it with simulated timeouts, 7.x pods, the domain 'example.com' and absolute nonsense hostnames - as a user the visible messages are OK for me. (And internal logs are more detailed).  You have the last word on this ;-)

jhass replied:
Well, summarizing everything under two user facing messages, one for timeouts and one for everything else, makes it pretty impossible to even begin to troubleshoot the error.

When there's bad network conditions, like the network dropped during a request, a hostname could not be resolved etc, this tells us nothing about the state of API support at the user's pod. And a bad network condition is nothing they need to bother their podmin about.

So improving their network connection is likely something the user can do something about, upgrading their pod likely is something they can't do anything about. It makes a lot of sense to me to try to separate these error classes, it will actually benefit the user a lot.

tclaus committed:
Update ios/Runner/AppAuthHandler.swift

Co-authored-by: Jonne Haß <me@jhass.eu>

tclaus committed:
Update ios/Runner/AppAuthHandler.swift

Co-authored-by: Jonne Haß <me@jhass.eu>

tclaus committed:
Update lib/src/localizations.dart

Co-authored-by: Jonne Haß <me@jhass.eu>

tclaus committed:
Update lib/src/app_auth.dart

Co-authored-by: Jonne Haß <me@jhass.eu>

tclaus committed:
Update lib/src/app_auth.dart

Co-authored-by: Jonne Haß <me@jhass.eu>

tclaus committed:
Update lib/sign_in_page.dart

Co-authored-by: Jonne Haß <me@jhass.eu>

tclaus commented:
(Code suggestions arte nice, but leads to too many extra commits, will collect them in one commit next time)

tclaus committed:
Update lib/src/localizations.dart

Co-authored-by: Jonne Haß <me@jhass.eu>

tclaus committed:
Update lib/src/localizations.dart

Co-authored-by: Jonne Haß <me@jhass.eu>

tclaus committed:
Update lib/sign_in_page.dart

Co-authored-by: Jonne Haß <me@jhass.eu>

tclaus committed:
Update lib/src/app_auth.dart

Co-authored-by: Jonne Haß <me@jhass.eu>

tclaus committed:
Update lib/src/app_auth.dart

Co-authored-by: Jonne Haß <me@jhass.eu>

tclaus committed:
Update lib/src/app_auth.dart

Co-authored-by: Jonne Haß <me@jhass.eu>

tclaus committed:
Update basic code style

jhass commented:
You can batch commit them from the files tab.

tclaus committed:
Remove exceptions from error text in signin

tclaus committed:
Added error details for details view in widget
